File: hostel_management_system_backend/Controllers/HostelVerificationController.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using hostel_management_system_backend.Exceptions;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

[ApiController]
[Route("api")]
public sealed class HostelVerificationController : ControllerBase
{
    private readonly IHostelVerificationService _verificationService;

    public HostelVerificationController(IHostelVerificationService verificationService)
    {
        _verificationService = verificationService;
    }

    [Authorize]
    [HttpPost("hostels/{hostelId:guid}/verification/request")]
    public async Task<ActionResult<HostelVerificationRequestReadDto>> RequestVerification(Guid hostelId, CancellationToken cancellationToken)
    {
        var ownerId = GetUserIdOrThrow();
        var created = await _verificationService.RequestVerificationAsync(hostelId, ownerId, cancellationToken);
        return Ok(created);
    }

    [Authorize]
    [HttpPost("verification-requests/{requestId:guid}/approve")]
    public async Task<ActionResult<HostelVerificationRequestReadDto>> Approve(Guid requestId, [FromBody] ReviewVerificationRequestDto dto, CancellationToken cancellationToken)
    {
        EnsureAdmin();
        var adminId = GetUserIdOrThrow();
        var approved = await _verificationService.ApproveVerificationAsync(requestId, adminId, dto.AdminNotes, cancellationToken);
        return Ok(approved);
    }

    [Authorize]
    [HttpPost("verification-requests/{requestId:guid}/reject")]
    public async Task<ActionResult<HostelVerificationRequestReadDto>> Reject(Guid requestId, [FromBody] ReviewVerificationRequestDto dto, CancellationToken cancellationToken)
    {
        EnsureAdmin();
        var adminId = GetUserIdOrThrow();
        var rejected = await _verificationService.RejectVerificationAsync(requestId, adminId, dto.AdminNotes, cancellationToken);
        return Ok(rejected);
    }

    [Authorize]
    [HttpGet("hostels/{hostelId:guid}/verification/requests")]
    public async Task<ActionResult<List<HostelVerificationRequestReadDto>>> GetForHostel(Guid hostelId, CancellationToken cancellationToken)
    {
        var requesterId = GetUserIdOrThrow();
        var isAdmin = User.IsInRole(UserRole.Admin.ToString());
        var requests = await _verificationService.GetForHostelAsync(hostelId, requesterId, isAdmin, cancellationToken);
        return Ok(requests);
    }

    private Guid GetUserIdOrThrow()
    {
        var sub = User.FindFirstValue(JwtRegisteredClaimNames.Sub)
            ?? User.FindFirstValue(ClaimTypes.NameIdentifier);
        if (!Guid.TryParse(sub, out var userId))
        {
            throw new UnauthorizedException("Invalid access token.", "invalid_token");
        }

        return userId;
    }

    private void EnsureAdmin()
    {
        if (!User.IsInRole(UserRole.Admin.ToString()))
        {
            throw new ForbiddenException("Only admins can review verification requests.", "admin_only");
        }
    }
}

File: src/hostels/verification/hostel-verification.controller.ts
import {
  Body,
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseUUIDPipe,
  Post,
  UseGuards,
} from '@nestjs/common';

import { CurrentUserDecorator } from '../../common/decorators/current-user.decorator';
import { UserRole } from '../../common/enums/app.enums';
import { AppForbiddenException } from '../../common/exceptions/app-exception';
import { JwtAuthGuard } from '../../common/guards/jwt-auth.guard';
import { CurrentUser } from '../../common/interfaces/current-user.interface';
import { HostelVerificationService } from './hostel-verification.service';
import { HostelVerificationRequestReadDto, ReviewVerificationRequestDto } from './dto/hostel-verification.dto';

@Controller()
export class HostelVerificationController {
  constructor(private readonly hostelVerificationService: HostelVerificationService) {}

  @Post('hostels/:hostelId/verification/request')
  @UseGuards(JwtAuthGuard)
  @HttpCode(HttpStatus.OK)
  requestVerification(
    @Param('hostelId', ParseUUIDPipe) hostelId: string,
    @CurrentUserDecorator() currentUser: CurrentUser,
  ): Promise<HostelVerificationRequestReadDto> {
    return this.hostelVerificationService.requestVerification(hostelId, currentUser.userId);
  }

  @Post('verification-requests/:requestId/approve')
  @UseGuards(JwtAuthGuard)
  @HttpCode(HttpStatus.OK)
  approve(
    @Param('requestId', ParseUUIDPipe) requestId: string,
    @CurrentUserDecorator() currentUser: CurrentUser,
    @Body() dto: ReviewVerificationRequestDto,
  ): Promise<HostelVerificationRequestReadDto> {
    if (currentUser.role !== UserRole.Admin) {
      throw new AppForbiddenException('Only admins can review verification requests.', 'admin_only');
    }

    return this.hostelVerificationService.approveVerification(requestId, currentUser.userId, dto.adminNotes);
  }

  @Post('verification-requests/:requestId/reject')
  @UseGuards(JwtAuthGuard)
  @HttpCode(HttpStatus.OK)
  reject(
    @Param('requestId', ParseUUIDPipe) requestId: string,
    @CurrentUserDecorator() currentUser: CurrentUser,
    @Body() dto: ReviewVerificationRequestDto,
  ): Promise<HostelVerificationRequestReadDto> {
    if (currentUser.role !== UserRole.Admin) {
      throw new AppForbiddenException('Only admins can review verification requests.', 'admin_only');
    }

    return this.hostelVerificationService.rejectVerification(requestId, currentUser.userId, dto.adminNotes);
  }

  @Get('hostels/:hostelId/verification/requests')
  @UseGuards(JwtAuthGuard)
  getForHostel(
    @Param('hostelId', ParseUUIDPipe) hostelId: string,
    @CurrentUserDecorator() currentUser: CurrentUser,
  ): Promise<HostelVerificationRequestReadDto[]> {
    return this.hostelVerificationService.getForHostel(
      hostelId,
      currentUser.userId,
      currentUser.role === UserRole.Admin,
    );
  }
}
